replica/table.cc

```diff
@@ -2942,25 +2958,33 @@ void table::set_schema(schema_ptr s) {
     tlogger.debug("Changing schema version of {}.{} ({}) from {} to {}",
                 _schema->ks_name(), _schema->cf_name(), _schema->id(), _schema->version(), s->version());
 
-    for_each_compaction_group([&] (compaction_group& cg) {
-        for (auto& m: *cg.memtables()) {
-            m->set_schema(s);
+    with_lock(_flush_timer_mutex, [this, &s] {
+        _flush_timer.cancel();
+
+        for_each_compaction_group([&] (compaction_group& cg) {
+            for (auto& m: *cg.memtables()) {
+                m->set_schema(s);
+            }
+        });
+
+        _cache.set_schema(s);
+        if (_counter_cell_locks) {
+            _counter_cell_locks->set_schema(s);
         }
-    });
+        _schema = std::move(s);
 
-    _cache.set_schema(s);
-    if (_counter_cell_locks) {
-        _counter_cell_locks->set_schema(s);
-    }
-    _schema = std::move(s);
+        for (auto&& v : _views) {
+            v->view_info()->set_base_info(
+                v->view_info()->make_base_dependent_view_info(*_schema));
+        }
 
-    for (auto&& v : _views) {
-        v->view_info()->set_base_info(
-            v->view_info()->make_base_dependent_view_info(*_schema));
-    }
+        set_compaction_strategy(_schema->compaction_strategy());
+        trigger_compaction();
 
-    set_compaction_strategy(_schema->compaction_strategy());
-    trigger_compaction();
+        if (_schema->memtable_flush_period() > 0) {
+            _flush_timer.arm(std::chrono::milliseconds(_schema->memtable_flush_period()));
+        }
+    }).get();
 }
```

Comment:
I think this is problematic since yielding here leads to atomicity loss when updating schema of base table and its views. Why do we need the mutex?

Reply:
I see it's for the race between set_schema() and the timer handler.

You can take advantage that set_schema() is atomic, it will be completed in one go. 

If set_schema() rearmed, and on_flush_timer() tries to rearm it again, it's not a problem since it will cancel current one and do it again to the next interval. 

I suggest you use rearm here in set_schema() too since you unconditionally canceled it first. This check '_schema->memtable_flush_period() > 0' should also be present in the handler, so if schema disables the period during the flush, we avoid the rearm on completion

Reply:
If you need to have some way to avoid reentrancy in the handler, you can use a semaphore that is used only in the handler.

Reply:
Btw, the reason of failure is the get() here, it fails on debug only because debug forces preemption and get() is only supposed to be used within a thread or futures known to be resolved, for example, inside then_wrapped().

Reply:
we can consider `set_schema` is atomic because it's not async function returning future and there are no seastar thread yielding inside?

Reply:
if so seems we do not need any mutex here (fortunately) and just use rearm timer instead of arm in both functions

Reply:
> we can consider `set_schema` is atomic because it's not async function returning future and there are no seastar thread yielding inside?

correct. all of it will happen in one go without any yielding. so the finally block in the timer handler and set_schema() are serialized. I think you can take advantage of that for handling the race. The cancel() and arm() in set_schema() will happen without any other seastar task running in between (includes timer handler which is possibly invoked only when seastar reactor has a chance to run).

Reply:
Thanks so much! It is not so clear to understand all of these things just reading seastar docs: they're nice but miss many important details. (

Reply:
> Thanks so much! It is not so clear to understand all of these things just reading seastar docs: they're nice but miss many important details. (

Indeed. Let's see what we can do to improve it. /cc @nyh

Reply:
@nyh looks like we have some room for adding a section that explains the atomicity of non-yielding functions, that even timer handler is also a seastar function and will only be invoked when the current task is preempted out. Maybe with a similar example as in this patch. It's easy for people not familiar with our model to easily assume things like timer is handled with a different linux thread, allowing it to run concurrently with seastar tasks.

Reply:
It woluld be nice to have some article explaining atomicity and preemption in seastar. With examples )